websockets/feeds/cinema.go

```diff
@@ -0,0 +1,315 @@
+package feeds
+
+import (
+	"github.com/bakape/meguca/common"
+	"github.com/bakape/meguca/util"
+	"time"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"sync"
+	"strings"
+	"strconv"
+
+	"github.com/go-playground/log"
+)
+
+type cinemaStatus struct {
+	watching int
+	playing string
+	mu sync.RWMutex
+}
+
+type cinemaFeed struct {
+	baseFeed
+	mainFeed *Feed
+	thread uint64
+	playlist []cinemaVideo
+	startTime time.Time
+	push chan string
+	voteSkip chan string
+	skippers map[string]struct{}
+	videoTimer <-chan time.Time
+	status cinemaStatus
+}
+
+type cinemaVideo struct {
+	Url string `json:"url"`
+	Duration int64 `json:"duration"`
+	Title string `json:"title"`
+	Type string `json:"type"`
+}
+
+type cinemaMessage struct {
+	Cmd  string      `json:"cmd"`
+	Data interface{} `json:"data"`
+}
+
+func PushToCinema(thread uint64, url string, ip string) {
+	if feed := feeds.cinemaFeeds[thread] ; feed != nil {
+		for c := range feed.clients {
+			if c.IP() == ip {
+				feed.push <- url
+				break
+			}
+		}
+	}
+}
+
+func CinemaVoteSkip(thread uint64, ip string) {
+	if feed := feeds.cinemaFeeds[thread] ; feed != nil {
+		feed.voteSkip <- ip
+	}
+}
+
+func newCinemaFeed(thread uint64, mainFeed *Feed) *cinemaFeed {
+	var cf = &cinemaFeed{}
+	cf.thread = thread
+	cf.mainFeed = mainFeed
+	cf.push = make(chan string)
+	cf.voteSkip = make(chan string)
+	cf.skippers = make(map[string]struct{})
+	cf.init()
+	return cf
+}
+
+func (f* cinemaFeed) getStatus() (int, string) {
+	f.status.mu.Lock()
+	defer f.status.mu.Unlock()
+	return f.status.watching, f.status.playing
+}
+
+func (f *cinemaFeed) start() (err error) {
+	go func() {
+		ticker := time.NewTicker(time.Second*5)
+		for {
+			select {
+			case c := <-f.add:
+				f.addClient(c)
+				c.Send(f.envelopPlaylist())
+				f.updateStatus()
+			case c := <-f.remove:
+				cIP := c.IP()
+				if f.removeClient(c) { // end yourself if last client left
+					f.playlist = f.playlist[:0]
+					f.updateStatus()
+					return
+				}
+				// update skippers list
+				if _, ok := f.skippers[cIP] ; ok {
+					ipActive := false
+					for otherC := range f.clients {
+						if cIP == otherC.IP() {
+							ipActive = true
+							break
+						}
+					}
+					if !ipActive {
+						delete(f.skippers, cIP)
+					}
+				}
+				// send status info
+				f.updateStatus()
+			case <-ticker.C:
+				if len(f.playlist) > 0 {
+					f.sendToAll(f.envelopSyncTime())
+				}
+			case url := <-f.push:
+				cv, err := parseUrl(url)
```

Comment:
This will block the entire thread's loop on external I/O - unacceptable, because this function can be triggered by anybody. Easy DoS. Please add a requirement to be the current board's staff to push videos.